File: Solutions/PAT/Advanced/1054.c
#include <stdio.h>
/* The Dominant Color (20) */

int main() {
  int n, m;
  scanf("%d%d", &m, &n);

  /* 
    O(N) algorithm for finding 
    dominant element in a data stream
  */
  int cur = -1;   // the consecutive elmt in current consideration
  int ans = -1;   // final result elmt
  int cnt = 0;    // count the presence of cur
  int maxcnt = 0; // count the presence of ans
  for (int i=0; i<n; i++)
    for (int j=0; j<m; j++) {
      int c;
      scanf("%d", &c);
      if (!i && !j) ans = c;  // init ans
      if (c == cur) {
        cnt++;
        if (cnt > maxcnt) { // MUST be strictly greater than '>'
          maxcnt = cnt;
          ans = cur;
        }
      } else cur = c;
    }
  
  printf("%d\n", ans);
}

File: Solutions/PAT/Advanced/1054.lua
#!/usr/bin/env lua5.3
-- The Dominant Color (20)

d = {}    -- plain algorithm, count presences (TLE)
m = io.read('*number')
n = io.read('*number')
for i = 1, n do
  for j = 1, m do
    c = io.read('*number')
    d[c] = (d[c] ~= nil) and  (d[c] + 1) or 1
  end
end

cnt = 0
ans = nil
for k, v in pairs(d) do
  if v > cnt then
    cnt = v
    ans = k
  end
end
print(ans)